### PR TITLE
Move changes from crypto3-zk up to commit a9b351be

### PR DESCRIPTION
**Pull request by martun**

### CODE REVIEW

File: include/nil/actor/zk/snark/arithmetization/plonk/constraint.hpp [hunk 1]
```diff
@@ -138,27 +134,21 @@ namespace nil {
...
                                 }
-
-                                term_value = term_value * assignment;
-                            }
-                            acc = acc + term_value;
-                        }
-                        return make_ready_future<math::polynomial<typename VariableType::assignment_type>>(acc);
+                                return assignment;
+                            });
+                        return evaluator.evaluate().get();
```
Comment (Zerg1996) [inline include/nil/actor/zk/snark/arithmetization/plonk/constraint.hpp:139]: Hm, I think I don't understand how it's work... 

Function return `future<math::polynomial<typename VariableType::assignment_type>>`
but we return `return evaluator.evaluate().get();` so, if we just return evaluator.evaluate() it's be already future? So we don't need to .get().

File: include/nil/actor/zk/snark/arithmetization/plonk/constraint.hpp [hunk 2]
```diff
@@ -138,27 +134,21 @@ namespace nil {
...
                     }
 
                     template<typename ArithmetizationParams>
-                    future<math::polynomial_dfs<typename VariableType::assignment_type>>
+                    math::polynomial_dfs<typename VariableType::assignment_type>
```
Comment (Zerg1996) [inline include/nil/actor/zk/snark/arithmetization/plonk/constraint.hpp:143]: And it's strange that we have two function `evaluate` but one of them return future, another - polynomial_dfs. But they both parallel.

File: test/systems/plonk/placeholder/placeholder.cpp [hunk 3]
```diff
@@ -266,8 +266,6 @@ ACTOR_THREAD_TEST_CASE(placeholder_permutation_polynomials_test) {
     BOOST_CHECK_MESSAGE(id_res == sigma_res, "Complex check");
 }
 
-/*
```
Comment (martun) [inline test/systems/plonk/placeholder/placeholder.cpp:269]: I enabled all the tests that were enabled on the single threaded version. There were 2 tests with *boost::unit_test::disabled(), which I commented out. These 2 are related to lookup argument, which will be updated, no need to fix it.

File: include/nil/actor/zk/snark/systems/plonk/placeholder/verifier.hpp [hunk 4]
```diff
@@ -88,13 +88,10 @@ namespace nil {
...
                         for (std::size_t i = 0; i < permutation_size; i++) {
                             std::size_t zero_index = 0;
-                            for (std::size_t j = 0;
-                                 j < preprocessed_public_data.common_data.columns_rotations[i].size();
-                                 j++) {
-
-                                if (preprocessed_public_data.common_data.columns_rotations[i][j] == 0) {
-                                    zero_index = j;
-                                }
+                            for (int v: preprocessed_public_data.common_data.columns_rotations[i]) {
```
Comment (Zerg1996) [inline include/nil/actor/zk/snark/systems/plonk/placeholder/verifier.hpp:91]: We need to do some refactoring `zk` library I guess...

```
std::size_t zero_index = std::find(preprocessed_public_data.common_data.columns_rotations[i].begin(), preprocessed_public_data.common_data.columns_rotations[i].end(), 0)
```

File: include/nil/actor/zk/snark/systems/plonk/placeholder/verifier.hpp [hunk 5]
```diff
@@ -118,59 +115,54 @@ namespace nil {
                         typename policy_type::evaluation_map columns_at_y;
                         for (std::size_t i = 0; i < witness_columns; i++) {
                             std::size_t i_global_index = i;
```
Comment (Zerg1996) [inline include/nil/actor/zk/snark/systems/plonk/placeholder/verifier.hpp:117]: Can you check please - is `i_global_index` and `i` it's always the same thing? So we don't need on of them.

File: test/systems/plonk/placeholder/placeholder.cpp [hunk 6]
```diff
@@ -343,115 +341,117 @@ ACTOR_THREAD_TEST_CASE(placeholder_permutation_argument_test) {
...
-
-    for (int i = 0; i < argument_size; i++) {
-        BOOST_CHECK(prover_res.F[i].evaluate(y) == verifier_res[i]);
-        for (std::size_t j = 0; j < desc.rows_amount; j++) {
-            BOOST_CHECK(prover_res.F[i].evaluate(preprocessed_public_data.common_data.basic_domain->get_domain_element(
-                            j)) == FieldType::value_type::zero());
-        }
-    }
-}
+// This test is disabled in singlethreaded version due to a bug in zk/snark/systems/plonk/placeholder/lookup_argument.hpp line 160-173.
```
Comment (Zerg1996) [inline test/systems/plonk/placeholder/placeholder.cpp:344]: Let's use ?
```
#ifdef 0
...
#endif
```

It's look little bit beatufily than many line of commendet code.

**Review by Zerg1996 — CHANGES_REQUESTED** [review]

File: include/nil/actor/zk/snark/systems/plonk/placeholder/verifier.hpp (hunk 4 above)
Comment (martun) [reply include/nil/actor/zk/snark/systems/plonk/placeholder/verifier.hpp:91]: That's true, but let's do it later. I'm planning to read the placeholder code and try to refactor/optimize it in a few weeks. There are plenty of such small things, and I'm removing them over time.

File: test/systems/plonk/placeholder/placeholder.cpp (hunk 6 above)
Comment (martun) [reply test/systems/plonk/placeholder/placeholder.cpp:344]: Apparently you can not ifdef 0, so I added #ifdef ENABLE_PLACEHOLDER_LOOKUP_TESTS

Comment (Zerg1996) [reply test/systems/plonk/placeholder/placeholder.cpp:344]: Oh, yeah, u right. 
```
#if 0
```

But anyway if we have something like #ifdef ENABLE_PLACEHOLDER_LOOKUP_TESTS it is much better!

**Review by Zerg1996 — APPROVED** [review]

**Pull request merged**